crates/bitwarden/src/platform/fido2/types.rs

```diff
@@ -0,0 +1,256 @@
+    pub pub_key_cred_params: Vec<PublicKeyCredentialParameters>,
+    pub exclude_list: Option<Vec<PublicKeyCredentialDescriptor>>,
+    pub options: Options,
+    pub extensions: Extensions,
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct MakeCredentialResult {
+    pub authenticator_data: Vec<u8>,
+    pub attested_credential_data: Vec<u8>,
+    pub credential_id: Vec<u8>,
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct GetAssertionRequest {
+    pub rp_id: String,
+    pub client_data_hash: Vec<u8>,
+    pub allow_list: Option<Vec<PublicKeyCredentialDescriptor>>,
+    pub options: Options,
+    pub extensions: Extensions,
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct Options {
+    pub rk: bool,
+    pub uv: UV,
+}
+
+impl From<super::CheckUserOptions> for Options {
+    fn from(value: super::CheckUserOptions) -> Self {
+        Self {
+            rk: value.require_presence,
+            uv: value.require_verification.into(),
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "mobile", derive(uniffi::Enum))]
+pub enum UV {
+    Discouraged,
+    Preferred,
+    Required,
+}
+
+impl From<UV> for Verification {
+    fn from(value: UV) -> Self {
+        match value {
+            UV::Discouraged => Verification::Discouraged,
+            UV::Preferred => Verification::Preferred,
+            UV::Required => Verification::Required,
+        }
+    }
+}
+
+impl From<Verification> for UV {
+    fn from(value: Verification) -> Self {
+        match value {
+            Verification::Discouraged => UV::Discouraged,
+            Verification::Preferred => UV::Preferred,
+            Verification::Required => UV::Required,
+        }
+    }
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct GetAssertionResult {
+    pub credential_id: Vec<u8>,
+    pub authenticator_data: Vec<u8>,
+    pub signature: Vec<u8>,
+    pub user_handle: Vec<u8>,
+    /**
+     * SDK IMPL NOTE: This is not part of the spec and is not returned by passkey-rs.
+     * The SDK needs to add this after the response from passkey-rs is received.
+     */
+    pub selected_credential: SelectedCredential,
+}
+
+#[cfg_attr(feature = "mobile", derive(uniffi::Enum))]
+pub enum ClientData {
+    DefaultWithExtraData { android_package_name: String },
+    DefaultWithCustomHash { hash: Vec<u8> },
+}
+
+// To keep compatibility with passkey-rs ClientData, we want extra_client_data to return a type
+// that serializes to Object | Unit. If we just used an Option it would serialize to Object | None
+// So we create our own wrapper types that serialize to the correct values and make the ClientData
+// implementation return that.
+#[derive(Clone)]
+pub(super) struct OptionalAndroidClientData {
+    data: Option<AndroidClientData>,
+}
+
+#[derive(Serialize, Clone)]
+struct AndroidClientData {
+    android_package_name: String,
+}
```

Comment:
question: Does the `AndroidClientData` struct need to be optional though? The functions that need it will never be used by anything other than Android?

Reply:
The value is optional because we need a generic type that implements serialize and covers both the `DefaultWithExtraData` and `DefaultWithCustomHash` branches of the enum.

This way, `extra_client_data` returns the following:

- For `DefaultWithExtraData` we return `Some(AndroidClientData)`, which gets serialized and then flattened, adding the `androidPackageName` field to `ClientDataJson`
- For `DefaultWithCustomHash` we return `None`, which gets serialized and then flattened, and leaves the `ClientDataJson` as-is


We could also change the type to be something like this, which might work as well:
``` 
#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct SerializableClientData {
    #[serde(skip_serializing_if = "Option::is_none")]
    android_package_name: Option<String>,
}
```